src/actions.ts

```diff
@@ -0,0 +1,449 @@
+		const server = await this.getServer(true);
+		if (!server) {
+			return {actionsuccess: false};
+		}
+
+		const date = parseInt(params.date);
+		const usercount = parseInt(params.users || params.usercount);
+		if (isNaN(date) || isNaN(usercount)) {
+			return {actionsuccess: false};
+		}
+
+		await tables.userstats.insert({
+			serverid: server.id, date, usercount,
+		}, SQL`ON DUPLICATE KEY UPDATE \`date\`= ${date}, \`usercount\`= ${usercount}`);
+
+		if (server.id === Config.mainserver) {
+			await tables.userstatshistory.insert({date, usercount});
+		}
+		return {actionsuccess: true};
+	},
+	async upkeep(params) {
+		const challengeprefix = this.verifyCrossDomainRequest();
+		const res: {[k: string]: any} = {};
+		const curuser = this.user;
+		let userid = '';
+		if (curuser.loggedin) {
+			res.username = curuser.name;
+			userid = curuser.id;
+		} else if (this.cookies.get('showdown_username')) {
+			res.username = this.cookies.get('showdown_username')!;
+			userid = toID(res.username);
+		}
+		if (userid !== '') {
+			const challengekeyid = !params.challengekeyid ? -1 : parseInt(params.challengekeyid);
+			const challenge = params.challstr || "";
+			res.assertion = await this.session.getAssertion(
+				userid, challengekeyid, curuser, challenge, challengeprefix
+			);
+		}
+		res.loggedin = curuser.loggedin;
+		return res;
+	},
+	async json() {
+		if (!Dispatcher.isJSON(this.request)) {
+			throw new ActionError("/api/json must use application/json requests");
+		}
+		let json;
+		try {
+			json = await Dispatcher.parseSentRequest(this.request);
+		} catch {
+			return [{actionerror: 'Malformed JSON sent.'}];
+		}
+		if (!json || !Array.isArray(json)) {
+			throw new ActionError(`Malformed JSON (send a JSON array in the 'json' property).`);
+		}
+
+		let serverid, servertoken;
+		for (const req of json) {
+			if (!serverid) serverid = req.serverid;
+			if (!servertoken) servertoken = req.servertoken;
+		}
+		const server = Dispatcher.servers[toID(serverid)];
+		if (json.length > 20) {
+			if (!server || server.token && server.token !== md5(servertoken)) {
+				throw new ActionError(`Only registered servers can send >20 requests at once.`);
+			}
+		}
+
+		const results = [];
+		for (const request of json) {
+			if (request.actionerror) continue;
+			if (!request.act) {
+				results.push({actionerror: 'Must send a request type.'});
+				continue;
+			}
+			const dispatcher = new Dispatcher(this.request, this.response, {
+				body: request,
+				act: request.act,
+			});
+			try {
+				const result = await dispatcher.executeActions();
+				results.push(result);
+			} catch (e) {
+				if (e instanceof ActionError) {
+					results.push({actionerror: e.message});
+					continue;
+				}
+				throw e;
+			}
+		}
+		return results;
+	},
+	async prepreplay(params) {
+		const server = await this.getServer(true);
+		if (!server) {
+			return {errorip: this.getIp()};
+		}
+		const extractedFormatId = /^([a-z0-9]+)-[0-9]+$/.exec((`${params.id}`));
```

Comment:
Whats with the extra set of parens and unnecessary template string?

Reply:
The parens are a typo. The template string is to ensure the input is a string (it could be a number as well).

Reply:
> (it could be a number as well)
It can't according to your types, which has `params: {[k: string]: string}`

Reply:
OK, fixed types to ensure everything is always a string.